src/Storages/ExternalTable/ExternalTableFactory.h

```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <boost/core/noncopyable.hpp>
+#include "Storages/ExternalTable/IExternalTable.h"
+#include "Storages/ExternalTable/ExternalTableSettings.h"
+
+namespace DB
+{
+
+/// Allows to create an IExternalTable by the name of they type.
+class ExternalTableFactory final : private boost::noncopyable
+{
+public:
+    static ExternalTableFactory & instance();
+
+    using Creator = std::function<IExternalTablePtr(const String & /*name*/, ExternalTableSettingsPtr /*settings*/)>;
+
+    IExternalTablePtr getExternalTable(const String & name, ExternalTableSettingsPtr settings) const;
+    void registerExternalTable(const String & type, Creator creator);
+
+private:
+    ExternalTableFactory();
+
+    std::unordered_map<std::string, Creator> creators;
```

Comment:
Do we need this factory or we can just use existing table engine factory ?

Reply:
I don't think there is an existing factory we can use directly for external table implementations, and probably it's not a good idea to mix things up. So having its own factory for external tables should make sense.

Reply:
Because of the column descriptions missing ?

Reply:
Let's make it another discussion and make another PR if needed.